<a href="https://colab.research.google.com/github/sebastianrdz/portafolio-implementacion/blob/main/A00827463_SebastianRdz_Modulo2_Modelo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Momento de Retroalimentación: Módulo 2 Uso de framework o biblioteca de aprendizaje máquina para la implementación de una solución. (Portafolio Implementación)

## Conexion Google Dirve & Redirecionamiento de Directorio

In [156]:
from google.colab import drive
drive.mount("/content/gdrive")  
!pwd

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).
/content/gdrive/My Drive/Colab Notebooks


In [157]:
%cd "/content/gdrive/MyDrive/Colab Notebooks" 
!ls

/content/gdrive/MyDrive/Colab Notebooks
 A00827463_SebastianRdz_Modulo1_Mercurio.ipynb		   mercurio.csv
 A00827463_SebastianRdz_Modulo1_Salarios.ipynb		   Mod2_Aug10.ipynb
 A00827463_SebastianRdz_Modulo2_Modelo.ipynb		   PlayDataset.csv
 A00827463_SebastianRdz_Modulo2_RegresionLineal.ipynb	   PlayDataset.ipynb
'Copia de A00827463_SebastianRdz_Modulo1_Salarios.ipynb'   Practicas
 ds_salaries.csv					   Retos
 mc-donalds-menu.csv					   wine.data


## Librerias

In [158]:
# Import required libraries
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt
import sklearn
from sklearn.neural_network import MLPClassifier
from sklearn.neural_network import MLPRegressor

# Import necessary modules
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from math import sqrt
from sklearn.metrics import r2_score
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import classification_report,confusion_matrix

# Manipulacion de datos

In [159]:
# Carga de de dataset
names = ["Class", "Alcohol", "Malic acid", "Ash", "Alcalinity of ash", "Magnesium", "Total phenols", "Flavanoids", "Nonflavanoid phenols", "Proanthocyanins", "Color intensity", "Hue", "OD280/OD315 of diluted wines", "Proline"]
df = pd.read_csv('wine.data', names=names)
# print(df.shape)
# df.describe().T
df.head()
# df.info()

Class  Alcohol  Malic acid   Ash  Alcalinity of ash  Magnesium  \
0      1    14.23        1.71  2.43               15.6        127   
1      1    13.20        1.78  2.14               11.2        100   
2      1    13.16        2.36  2.67               18.6        101   
3      1    14.37        1.95  2.50               16.8        113   
4      1    13.24        2.59  2.87               21.0        118   

   Total phenols  Flavanoids  Nonflavanoid phenols  Proanthocyanins  \
0           2.80        3.06                  0.28             2.29   
1           2.65        2.76                  0.26             1.28   
2           2.80        3.24                  0.30             2.81   
3           3.85        3.49                  0.24             2.18   
4           2.80        2.69                  0.39             1.82   

   Color intensity   Hue  OD280/OD315 of diluted wines  Proline  
0             5.64  1.04                          3.92     1065  
1             4.38  1.05                          3.40     1050  
2             5.68  1.03                          3.17     1185  
3             7.80  0.86                          3.45     1480  
4             4.32  1.04                          2.93      735

In [160]:
# Separamos la variable a analizar, en este caso la classe a la que pertence el alchol.
# Tomamos las columnas resultantes para considerrals dentro de nuestro modelo y esablecemos 
# sus valores con base a 1 para poder visualizar los resultados con base a un total de 1.00 (100%)
y = ['Class'] 
x = list(set(list(df.columns))-set(y))
df[x] = df[x]/df[x].max()
df.describe().T

count      mean       std       min       25%  \
Class                         178.0  1.938202  0.775035  1.000000  1.000000   
Alcohol                       178.0  0.876643  0.054742  0.743763  0.833614   
Malic acid                    178.0  0.402819  0.192611  0.127586  0.276293   
Ash                           178.0  0.732668  0.084936  0.421053  0.684211   
Alcalinity of ash             178.0  0.649831  0.111319  0.353333  0.573333   
Magnesium                     178.0  0.615689  0.088163  0.432099  0.543210   
Total phenols                 178.0  0.591524  0.161302  0.252577  0.449098   
Flavanoids                    178.0  0.399463  0.196626  0.066929  0.237205   
Nonflavanoid phenols          178.0  0.548264  0.188566  0.196970  0.409091   
Proanthocyanins               178.0  0.444385  0.159877  0.114525  0.349162   
Color intensity               178.0  0.389084  0.178330  0.098462  0.247692   
Hue                           178.0  0.559912  0.133668  0.280702  0.457602   
OD280/OD315 of diluted wines  178.0  0.652921  0.177498  0.317500  0.484375   
Proline                       178.0  0.444579  0.187445  0.165476  0.297917   

                                   50%       75%  max  
Class                         2.000000  3.000000  3.0  
Alcohol                       0.879973  0.922286  1.0  
Malic acid                    0.321552  0.531466  1.0  
Ash                           0.730650  0.791796  1.0  
Alcalinity of ash             0.650000  0.716667  1.0  
Magnesium                     0.604938  0.660494  1.0  
Total phenols                 0.606959  0.721649  1.0  
Flavanoids                    0.420276  0.565945  1.0  
Nonflavanoid phenols          0.515152  0.662879  1.0  
Proanthocyanins               0.434358  0.544693  1.0  
Color intensity               0.360769  0.476923  1.0  
Hue                           0.564327  0.654971  1.0  
OD280/OD315 of diluted wines  0.695000  0.792500  1.0  
Proline                       0.400893  0.586310  1.0

In [161]:
# Tomamos los valores de los datos anteriores y utilizamos esto para separar en grupos que podamos 
# analizar para entrenar, y otro para poder probar nuestro modelo.
X_train, X_test, y_train, y_test = train_test_split(df[x].values, df[y].values, test_size=0.2, random_state=100)
print(X_train.shape)
print(X_test.shape)

(142, 13)
(36, 13)


In [162]:
# Definimos el modelo a ulilizar (Neural Network Multi-layer Perceptron classifier) y lo inicializamos.
# Procedente, utilizamos el modelo y nuestros datos para entrenar con los datos de train y la funcion fit.
# Con base a este modelo ya entrenado, podemos predecir con los valores de x, la classe a la que el alcol pertenece.
mlp = MLPClassifier(hidden_layer_sizes=(13,13,13), activation='relu', solver='adam', max_iter=1000)
mlp.fit(X_train,y_train)

predict_train = mlp.predict(X_train)
predict_test = mlp.predict(X_test)

In [163]:
# Utilizamos "confusion_matrix" para representar como es que se destribuyen los datos atraves de las 3 classes que tenemos precentes.
# Utilizamos "classification_report" para representar la correlaciones y el grado de confiabilidad del modelo con base a los datos proporcionados.
# Todo esto con base a los datos de entrenamiento.
print(confusion_matrix(y_train, predict_train))
print(classification_report(y_train, predict_train))

[[51  0  0]
 [ 0 56  0]
 [ 0  0 35]]
              precision    recall  f1-score   support

           1       1.00      1.00      1.00        51
           2       1.00      1.00      1.00        56
           3       1.00      1.00      1.00        35

    accuracy                           1.00       142
   macro avg       1.00      1.00      1.00       142
weighted avg       1.00      1.00      1.00       142



In [164]:
# Aqui podemos ver las prediciones con base a nuestro modelo y nuestros datos de testing.
print(confusion_matrix(y_test, predict_test))
print(classification_report(y_test, predict_test))

[[ 8  0  0]
 [ 0 15  0]
 [ 0  1 12]]
              precision    recall  f1-score   support

           1       1.00      1.00      1.00         8
           2       0.94      1.00      0.97        15
           3       1.00      0.92      0.96        13

    accuracy                           0.97        36
   macro avg       0.98      0.97      0.98        36
weighted avg       0.97      0.97      0.97        36



## Conclusión
Dentro de Este modelo se implemento la librería Sklearn para para llevar acabo el modelo nombrado _Neural Network Multi-layer Perceptron classifier_, el cual nos permite modelar con un formato a e conexiones neuronales para hacer predicciones.
<br>

Así mismo, se utilizo el data set ya definido en este curso de _wines.csv_ para correr las predicciones dentro de este. Como métrica de desempeño se considero la clase a la que pertenece cada alcohol (1, 2 o 3). 
<br>

Para definir los valores de entrenamiento, y los valores de prueba, utilize una librería igualmente de Sklearn para dividir los datos de manera aleatoria. Esto nos da nuestros valores Y y X de entrenamiento al igual que nuestros valores Y y X de prueba. Teniendo nuestros valores inicializamos el modelo ya definido por Sklearn, y lo entrenamos por medio de la función _fit()_ en donde le pasamos nuestros valores X de entrenamiento y prueba. Teniendo esto hacemos una predicción con base a nuestras classes para ajustar el modelo con base a las X.
<br>

Teniendo esta predicción acabamos! Podemos visualizar los resultados de dos maneras:
- "confusion_matrix" para representar cómo es que se distribuyen los datos través de las 3 classes que tenemos presentes.
- "classification_report" para representar la correlaciones y el grado de confiabilidad del modelo con base a los datos proporcionados.

En las pruebas realizadas arriba, se puede visualizar que con los datos de entrenamiento obtuvimos un grado de confianza y certeza de 1.00 (100%)
Si embargo en la pruebas realizadas con los datos de prueba, obtuvimos un 0.97 (97%) de certeza.
<br>

Dicho esto, una limitación de esta prueba considero es el tamaño de la muestra. Wines tiene un conjunto de 178 campos por columna. Teniendo mas datos a los que analizar se pudiera realizar un mejor ajuste del modelo y posiblemente una mejor predicción al final.
